### ISSUE TITLE
Parser in domish does not keep unreferenced namespace declarations

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11177]
|<img alt="ff's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| ff reported|
|-|-|
|Trac ID|trac#1669|
|Type|defect|
|Created|2006-04-24 17:43:04Z|

When parsing xml chunks like this one

```
<SOAP-ENV:Envelope xmlns:SOAP-ENC='http://schemas.xmlsoap.org/soap/encoding/'
  xmlns:SOAP-ENV='http://schemas.xmlsoap.org/soap/envelope/'
  xmlns:xsd='http://www.w3.org/1999/XMLSchema'
  xmlns:xsi='http://www.w3.org/1999/XMLSchema-instance'
  SOAP-ENV:encodingStyle='http://schemas.xmlsoap.org/soap/encoding/'>
<SOAP-ENV:Body>
<TheMethod SOAP-ENC:root='1'>
<v1 xsi:type='xsd:string'>p1</v1>
<v2 xsi:type='xsd:string'>p2</v2>
</TheMethod>
</SOAP-ENV:Body>
</SOAP-ENV:Envelope>
```

and then re-serializing it, the parser in domish produces an output like the following one:
```
<xn0:Envelope xmlns:xn0='http://schemas.xmlsoap.org/soap/envelope/'
 xn0:encodingStyle='http://schemas.xmlsoap.org/soap/encoding/'>
<xn0:Body>
<TheMethod xmlns:xn1='http://schemas.xmlsoap.org/soap/encoding/' 
 xn1:root='1'>
<v1 xmlns:xn2='http://www.w3.org/1999/XMLSchema-instance'
 xn2:type='xsd:string'>p1</v1>
<v2 xmlns:xn2='http://www.w3.org/1999/XMLSchema-instance'
 xn2:type='xsd:string'>p2</v2>
</TheMethod>
</xn0:Body>
</xn0:Envelope>
```

In the result the declaration xmlns:xsd='http://www.w3.org/1999/XMLSchema' is no more reported, because it is not explicitely used in the document, but as value of xsi:type attribute. The usage of qualifying namespaces in attribute values is [questioned](http://www.xml.com/pub/a/2002/02/06/deviant.html), but used in SOAP, WSDL, XSLT and XML Schemas. For this reason it is important to keep track of namespace declarations where they are found, even if not used in the document.

A possibile solution could be to store namespace declarations as regular attributes of the elements where they appear.
 

<details><summary>Searchable metadata</summary>

```
trac-id__1669 1669
type__defect defect
reporter__ff ff
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__words words
keywords__ 
time__1145900584000000 1145900584000000
changetime__1146498021000000 1146498021000000
version__2_0 2.0
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)|@ralphm set owner to @ralphm<br>@ralphm set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)|@ralphm removed owner<br>@ralphm set status to `new`|
|-|-|

This is fixed in [branches/domish-keep-prefixes-1669.](branches/domish-keep-prefixes-1669.) Also, I fixed `DomishStreamTests` to not use `xpath`, but interact with the `Element` objects directly.

**Comment by twisted-trac** [comment]
|<img alt="tld's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|tld commented|
|-|-|

Code seems fine to me.  Could use more test cases, but then again, I always think code could use more test cases.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)|@ralphm set status to `closed`|
|-|-|

(In [16718]) Local prefix storage

Author: ralphm
Reviewer: tld
Fixes [#11177](https://github.com/twisted/twisted/issues/11177)

This allows for elements to store namespace declarations. When parsing and
subsequential serialization, the result should have the same prefixes defined
in the same place in the element tree.

  * Fix a parsing issue in the expat based parser so that the default
    namespace is '', not None for prefixed elements without a bare xmlns
    declaration.
  * Redo DomishStreamTests to not use xpath, but interact with the Element
    objects themselves.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
